### PR TITLE
provider/docker: Docker upload

### PR DESCRIPTION
**Pull request by ColinHebert**
This should allow the user to upload custom files/folders to the docker container before starting it.

In terms of command line, this is the equivalent of:
```bash
docker create --name blah image
docker cp local_path blah:remote_path
docker start blah
```

### CODE REVIEW

File: builtin/providers/docker/resource_docker_container.go [hunk 1]
```diff
@@ -2,10 +2,14 @@ package docker
...
 	"bytes"
+	"crypto/sha1"
+	"encoding/hex"
 	"fmt"
-
+	"io"
+	"os"
 	"regexp"
 
+	"github.com/fsouza/go-dockerclient/external/github.com/docker/docker/pkg/archive"
```
Comment (ColinHebert) [inline builtin/providers/docker/resource_docker_container.go]: Not sure whether transitive dependencies are the way to go here. Any suggestions?

File: builtin/providers/docker/resource_docker_container.go [hunk 2]
```diff
@@ -341,6 +345,54 @@ func resourceDockerContainer() *schema.Resource {
 				Elem:     &schema.Schema{Type: schema.TypeString},
 				Set:      stringSetHash,
 			},
+
+			"uploads": &schema.Schema{
```
Comment (ColinHebert) [inline builtin/providers/docker/resource_docker_container.go]: Could be renamed if necessary

File: builtin/providers/docker/resource_docker_container.go [hunk 3]
```diff
@@ -341,6 +345,54 @@ func resourceDockerContainer() *schema.Resource {
...
+							ForceNew: true,
+							ValidateFunc: func(v interface{}, k string) (ws []string, es []error) {
+								path := v.(string)
+								if _, err := os.Stat(path); err != nil {
+									es = append(es, fmt.Errorf(
+										"%q must be valid path: %s", path, err))
+								}
+								return
+							},
+							StateFunc: func(v interface{}) string {
```
Comment (ColinHebert) [inline builtin/providers/docker/resource_docker_container.go]: I'm not sure why but during my tests, it turns out that the generated state file leaves the path here but while doing a plan, the checksum is generated. This leads to changes such as

```
    uploads.#:                         "1" => "1"
    uploads.241032095.local_path:      "files/" => ""
    uploads.241032095.remote_path:     "/tmp/" => ""
    uploads.4134509434.local_path:     "" => "7c95cb15f351188c8a680e2a2e15df87d163cc8a" (forces new resource)
    uploads.4134509434.remote_path:    "" => "/tmp/" (forces new resource)
```

Could this be related to the many issues related to `StateFunc` that cropped up everywhere :(

File: builtin/providers/docker/resource_docker_container.go [hunk 4]
```diff
@@ -341,6 +345,54 @@ func resourceDockerContainer() *schema.Resource {
...
+								switch v.(type) {
+								case string:
+									reader, err := archive.Tar(v.(string), archive.Uncompressed)
+									if err != nil {
+										return "invalid"
```
Comment (ColinHebert) [inline builtin/providers/docker/resource_docker_container.go]: There is no nice error management in `StateFunc`, shame :(

**Comment by ColinHebert** [comment]
@jen20 @radeksimko, any opinion/ advices regarding the comments above?

**Comment by radeksimko** [comment]
Hi @ColinHebert 
Would you mind describing some use cases for this?

It feels like a functionality that belongs into either Packer or some kind of provisioner rather than a resource - but I _may_ misunderstood your intentions here.

**Comment by radeksimko** [comment]
I just found https://github.com/hashicorp/terraform/pull/6827 (docker provisioner) - would that be sufficient for your use case (when finished)?

**Comment by stack72** [comment]
@ColinHebert I am going to close this PR out. It's been a while since any communication on it - I am going to assume (for now) that it's not needed. If you feel it still is, please just have a read back through and let us know :)

Thanks for all the work here!

Paul

**Pull request closed without merging**